src/protocols/hpr_protocol_router.erl

```diff
@@ -98,14 +113,71 @@ get_stream(Gateway, LNS, Server) ->
                             Error;
                         {ok, Stream} ->
                             true = ets:insert(?STREAM_ETS, {{Gateway, LNS}, Stream}),
+                            ok = gen_server:cast(?MODULE, {monitor, Gateway, LNS}),
                             get_stream(Gateway, LNS, Server)
                     end
             end
     end.
 
--spec remove_stream(libp2p_crypto:pubkey_bin(), binary()) -> true.
+-spec remove_stream(libp2p_crypto:pubkey_bin(), binary()) -> ok.
 remove_stream(Gateway, LNS) ->
-    ets:delete(?STREAM_ETS, {Gateway, LNS}).
+    case ets:lookup(?STREAM_ETS, {Gateway, LNS}) of
+        [{_, Stream}] ->
+            ok = grpcbox_client:close_send(Stream),
+            true = ets:delete(?STREAM_ETS, {Gateway, LNS}),
+            lager:debug("closed stream"),
+            ok;
+        [] ->
+            lager:debug("did not close stream"),
+            ok
+    end.
+
+%% ------------------------------------------------------------------
+%% gen_server Function Definitions
+%% ------------------------------------------------------------------
+
+init(_Args) ->
+    {ok, #state{}}.
+
+-spec handle_call(Msg, _From, #state{}) -> {stop, {unimplemented_call, Msg}, #state{}}.
+handle_call(Msg, _From, State) ->
+    {stop, {unimplemented_call, Msg}, State}.
+
+handle_cast({monitor, Gateway, LNS}, State) ->
+    {ok, Pid} = hpr_packet_router_service:locate(Gateway),
+    _ = erlang:monitor(process, Pid, [{tag, {'DOWN', Gateway, LNS}}]),
+    lager:debug("monitoring gateway stream ~p", [Pid]),
+    {noreply, State};
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(
+    {{'DOWN', Gateway, LNS}, _Monitor, process, _Pid, _ExitReason},
+    #state{} = State
+) ->
+    lager:debug("gateway stream ~p went down: ~p waiting ~wms", [_Pid, _ExitReason, ?CLEANUP_TIME]),
+    erlang:spawn(fun() ->
+        timer:sleep(?CLEANUP_TIME),
+        case hpr_packet_router_service:locate(Gateway) of
+            {error, _Reason} ->
+                lager:debug("connot find a gateway stream: ~p, shutting down", [_Reason]),
+                ?MODULE:remove_stream(Gateway, LNS);
+            {ok, StreamPid} ->
```

Comment:
monitor the new stream pid, like in gwmp_worker?

Reply:
When a new stream is created it gets monitored every time look at line 116